### PR TITLE
Issue #1738: Changed unit_recall dialogue box 

### PR DESCRIPTION
**Pull request by universecoder**
Changed unit_recall dialogue box to sort by
1. Level
2. Experience to advance

### CODE REVIEW

File: src/gui/dialogs/unit_recall.cpp [hunk 1]
```diff
@@ -181,6 +181,13 @@ void unit_recall::pre_show(window& window)
 		find_widget<button>(&window, "show_help", false),
 		std::bind(&unit_recall::show_help, this, std::ref(window)));
 
+	list.register_sorting_option(0, [this](const int i) { return recall_list_[i]->level(); });
```
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: You're duplicating the registration of sorting options. Please don't do that: only edit the existing sorting lambdas in lines 240-246.

File: src/gui/dialogs/unit_recall.cpp [hunk 2]
```diff
@@ -240,7 +247,7 @@ void unit_recall::pre_show(window& window)
 	list.register_sorting_option(0, [this](const int i) { return recall_list_[i]->type_name().str(); });
 	list.register_sorting_option(1, [this](const int i) { return recall_list_[i]->name().str(); });
 	list.register_sorting_option(2, [this](const int i) { return recall_list_[i]->level(); });
-	list.register_sorting_option(3, [this](const int i) { return recall_list_[i]->experience(); });
+	list.register_sorting_option(3, [this](const int i) { return recall_list_[i]->experience_to_advance(); });
```
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: Please don't change how sorting by XP works. (This change is completely orthogonal to the target of sorting by both level and XP, as I explained.)

File: src/gui/dialogs/unit_recall.cpp [hunk 3]
```diff
@@ -146,6 +146,14 @@ static std::string get_title_suffix(int side_num)
 	return msg.str();
 }
 
+void unit_recall:unit_recall_default_compare(unit &first, unit &second) {
```
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: Existing practice in Wesnoth codebase is to attach the `&` to the type name: `unit& first, unit& second`

Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: Also, the opening brace of a function goes to a line of its own.

File: src/gui/dialogs/unit_recall.cpp [hunk 4]
```diff
@@ -146,6 +146,14 @@ static std::string get_title_suffix(int side_num)
...
+	if ( first.level() > second.level() ) return true;
```
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: Wesnoth codebase practice is to omit space between keywords and opening parentheses, as well as within parentheses.
```
if(first.level() > second.level()) return true;
```

File: src/gui/dialogs/unit_recall.cpp (hunk 3 above)
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: Oh, because you aren't mutating the units, it's better to receive them as const references: `const unit& first, const unit& second`

**Comment by gfgtdf** [comment]
this just changes the defualt order, it doesnt fix the level colummn, the expected behaviour that is aslo in 1.12 is: 
1) the Level sort column uses experience_to_advance as a secodn criteria
2) The dialogs starts by beeing sorted with the 'level' column.

Also youtr sort function is wrong, you need to change for `first->level() < second->level()` before chacking for experience_to_advance.

**Comment by jyrkive** [comment]
@gfgtdf is correct. The sort function is broken. If `first->level() < second->level()`, the function should return `false`, instead of proceeding to compare remaining XP to next level.

In my test, attempting to view the recall list crashes the game, because Microsoft's debug CRT checks that the sort function is valid. It calls the function a second time with the parameters swapped: it's an error if **both** calls return `true`.

In addition, as @gfgtdf says, it would be nice if the player would be able to sort the units in some other way, and then **return** to sorting them primarily by level and secondarily by remaining XP. And that would work if you simply changed the lambda in [unit_recall.cpp:242](https://github.com/wesnoth/wesnoth/blob/bb9ee94e4b8008f9988ea78ebf18ac974966c406/src/gui/dialogs/unit_recall.cpp#L242) to return a value that depends on both criteria. It would be literally a one-line change.

File: src/gui/dialogs/unit_recall.cpp [hunk 5]
```diff
@@ -248,7 +249,7 @@ void unit_recall::pre_show(window& window)
 
 	list.register_sorting_option(0, [this](const int i) { return recall_list_[i]->type_name().str(); });
 	list.register_sorting_option(1, [this](const int i) { return recall_list_[i]->name().str(); });
-	list.register_sorting_option(2, [this](const int i) { return recall_list_[i]->level(); });
+    list.register_sorting_option(2, [this](const int i) { return recall_list_[i]->level()*1000 - recall_list_[i]->experience_to_advance(); });
```
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: You changed the line to be indented by spaces. Please change it back to tab indentation.

File: src/gui/dialogs/unit_recall.cpp [hunk 6]
```diff
@@ -146,6 +146,14 @@ static std::string get_title_suffix(int side_num)
 	return msg.str();
 }
 
+bool unit_recall::unit_recall_default_compare(const unit_const_ptr first, const unit_const_ptr second)
+{
+	if (first->level() > second->level()) return true;
```
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: Your custom sorting code is redundant. The listbox sorts the units by level *by default*, which overrides the order you set here. Please remove the custom sorting code, and only retain the change to the lambda in line 242.

**Comment by jyrkive** [comment]
So, it turned out that your sorting lambda has a bug - an integer underflow.

If the unit is at level 0, the subtraction will yield a negative value. Unfortunately, `unit::experience_to_advance()` returns an unsigned integer, and C++ rules state that a binary operation where one operand is signed and the other is unsigned returns an unsigned value. Yes, even if the operation is subtraction.

Thus, instead of a negative value (-23 in my test case), the lambda returns an extremely large positive value. As a result, level 0 units are sorted to the **top** of the list instead of the bottom.

To fix it, convert the experience value to a signed integer before subtraction:

    list.register_sorting_option(2, [this](const int i) { return recall_list_[i]->level()*1000 - static_cast<int>(recall_list_[i]->experience_to_advance()); });

File: src/gui/dialogs/unit_recall.cpp [hunk 7]
```diff
@@ -239,7 +249,7 @@ void unit_recall::pre_show(window& window)
 
 	list.register_sorting_option(0, [this](const int i) { return recall_list_[i]->type_name().str(); });
 	list.register_sorting_option(1, [this](const int i) { return recall_list_[i]->name().str(); });
-	list.register_sorting_option(2, [this](const int i) { return recall_list_[i]->level(); });
+    list.register_sorting_option(2, [this](const int i) { return recall_list_[i]->level()*1000 - static_cast<int>(recall_list_[i]->experience_to_advance()); });
```
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: You lost the tab indentation at some point.

**Comment by gfgtdf** [comment]
that dialog is alrady by default sorted by column 2 (level, see https://github.com/wesnoth/wesnoth/blob/1.13.8/src/gui/dialogs/unit_recall.cpp#L248 ) so the `std::sort` line shoudl be unneeded.

**Comment by universecoder** [comment]
No, that line is only used to set the sorting option so that the user can
choose the desired option. There is no 'default' sort.

**Comment by gfgtdf** [comment]
I just tested it in 1.13.8 and there the recall list is intilially ordered by level when you open that dialog, so i'm quite sure that the std::sort line is not needed.

**Comment by jyrkive** [comment]
The difference that line gives is what happens when you open the recall dialog *and click the level header*. In that situation, the listbox goes to unsorted state, where it shows the units *in the order they were added*.

(Wesnoth listboxes are weird: they allow the player to view the items in the order they were added, even after sorting them by a certain criterion.)

**Comment by gfgtdf** [comment]
> the listbox goes to unsorted state, where it shows the units in the order they were added.

This is afaik the epected behviour that is also in 1.12

**Comment by jyrkive** [comment]
So, you think that the player should have the option to view the recall list in an essentially random order?

**Comment by gfgtdf** [comment]
At least that's what 1.12 does. I didn't play seroiusly for quite some time now so i cannot tell whether there are usecases for it, but assuming that the 'random' order is the order in which they are originally recruited i wouldn't say it is impossible..
